### PR TITLE
Implemented inverse hyperbolic asech class

### PR DESCRIPTION
**Pull request by aktech**
This PR implements the inverse hyperbolic secant class.

**References:**

[1] http://en.wikipedia.org/wiki/Hyperbolic_function
[2] http://dlmf.nist.gov/4.37
[3] http://functions.wolfram.com/ElementaryFunctions/ArcSech/

@jksuom @smichr @asmeurer

### CODE REVIEW

**Comment by jksuom** [comment]
There is a large number of specific values. I am not sure if so many would be needed, but probably they do no harm. However, testing the table values should be carefully planned. Even if one is willing to trust the source (Wolfram, it seems), there remains the possibility of a misprint, either in the source or in the copy. (E.g. ``asech(zoo) = i`` in Wolfram looks like one.) Asserting ``asech(key) == value`` does not help.

An additional inconvenience is that the branch cut covers most of the real line. The principal values of ``asec(x)`` are unambiguous only for ``x`` between 0 and 1. For other real arguments there are two boundary values. They should be chosen consistently, not only in relation to each other but also to other functions.

I don't think it would be necessary to test all specific values, but several tests could be based on the definition of ``asech`` as the inverse of ``sech``. Also identities expressing ``asech`` in terms of functions like ``acosh`` and ``asec`` could be used.

**Comment by aktech** [comment]
> There is a large number of specific values. I am not sure if so many would be needed, but probably they do no harm.

Since, in SymPy we are more of interested in symbolic results, I think it's a good idea to have symbolic output for atleast the general formulas available. <br/>
For example, If we don't have `(-sqrt(5) + 1)` in the dictionary of the value-key pair, we would get this:
```
In [3]: asech(-sqrt(5) + 1)
Out[3]: asech(-sqrt(5) + 1)

In [4]: asech(-sqrt(5) + 1).n()
Out[4]: 2.51327412287183*I
```

and, I think the following is preferred: 
```python
In [2]: asech(-sqrt(5) + 1)
Out[2]: 4*I*pi/5
```

> However, testing the table values should be carefully planned. Even if one is willing to trust the source (Wolfram, it seems), there remains the possibility of a misprint, either in the source or in the copy. (E.g. asech(zoo) = i in Wolfram looks like one. Asserting asech(key) == value does not help.

Though in this PR it returns [NaN](https://github.com/sympy/sympy/pull/9291/files#diff-fe6e1f3f7df16e25f26b46a856dc551eR1204).
Is there a better way for testing?

> I don't think it would be necessary to test all specific values,

That is done to increase coverage.

I have added tests for inverse and differentiation as well, in lines of the previous tests.

The code and tests written here for `asech` class is tried to be consistent with the code for other inverse hyperbolic functions (*See other inverse hyperbolic classes in the same file*). If there is some inconsistency, then all the other classes (inverse hyperbolic) should be modified as well, and probably that's beyond the scope of this PR.

**Comment by jksuom** [comment]
I am afraid that my last paragraph was too cryptic. What I wanted to say was that the test for a specific argument value, say, ``alpha``, could be of the type ``assert sech(asech(alpha)) == alpha`` (possibly with suitable simplification). This would explicitly verify that ``asech`` really is the inverse of ``sech``.  Similarly, for some other specific argument ``beta`` one could have an expression involving ``beta`` and ``acosh`` or ``asec`` on the right hand side. Such tests would tie up specific values of ``asech`` to those of the related functions.

A line of the type ``asech(alpha) == gamma`` will verify the code reading the table, but it does nothing to prove that the value really is ``gamma``. It should not be necessary to set up such a test for each entry in the table.

**Comment by aktech** [comment]
I totally agree with your suggestions, the only thing I would like to point out is that we should definitely add those tests mentioned by you, but the only difference is that we have to test them numerically (there is surely no harm in that).
Since we can't simply rely on `simplify()` every time, as in this example:

```python
In [10]: a = sech(asech(sqrt(2) - sqrt(6))).simplify()

In [11]: a
Out[11]: -4/(sqrt(2) + sqrt(6))

In [12]: a == sqrt(2) - sqrt(6)
Out[12]: False
```
Simplification can't be trusted always.
Though both of these `sqrt(2) - sqrt(6)` & `-4/(sqrt(2) + sqrt(6))` are equal.

Also in this:
```python
In [2]: asech(-sqrt(2) + sqrt(6))
Out[2]: I*pi/12

In [7]: acosh(1/(-sqrt(2) + sqrt(6)))
Out[7]: acosh(1/(-sqrt(2) + sqrt(6)))
```
Though testing numerically works well:

```python
In [4]: a = asech(sqrt(6) - sqrt(2)).n()
In [5]: a
Out[5]: 0.261799387799149*I

In [6]: b = (acosh(1/(sqrt(6) - sqrt(2)))).n()
In [7]: b
Out[6]: 0.261799387799149*I

In [10]: a == b
Out[10]: True
```

**Comment by aktech** [comment]
> It should not be necessary to set up such a test for each entry in the table.

The tests for each entry in the table is written so as to prevent any tampering in key-value pairs, i.e if someone accidentally edits those, then the tests must Fail.

**Comment by aktech** [comment]
I have Added Tests for specific values of `asech` to those of the related functions.

**Comment by jksuom** [comment]
I was, in fact, expecting that a straightforward simplify (with its current capabilities) would not suffice, and certainly not in all cases. But there might still be working variants, e.g., along the following lines
```
a = sech(asech(sqrt(2) - sqrt(6))).simplify()
((sqrt(2) - sqrt(6))/a).simplify() == 1
```
(with suitable simplify()'s.)

I have not studied the specific values of ``acosh`` and ``acos`` (as defined in SymPy) but there might exist some that would be in line with those of ``asech``. (With inverted argument in ``acosh`` and value multiplied by ``I`` in ``acos``, if I remember correctly...)

I should prefer several tests of the above types. They would also guard against inadvertent changes in the tables.

**Comment by aktech** [comment]
I have Added tests for properties of `asech` mentioned by you.
Thanks!

File: sympy/functions/elementary/tests/test_hyperbolic.py [hunk 1]
```diff
@@ -502,6 +502,66 @@ def test_acosh_series():
...
+
+    # sech(asech(x)) / x == 1
+    assert (sech(asech(sqrt(6) - sqrt(2))) / (sqrt(6) - sqrt(2))).simplify() == 1
+    assert (sech(asech(sqrt(6) + sqrt(2))) / (sqrt(6) + sqrt(2))).simplify() == 1
+    assert (sech(asech(sqrt(2 + 2/sqrt(5)))) / (sqrt(2 + 2/sqrt(5)))).simplify() == 1
+    assert (sech(asech(-sqrt(2 + 2/sqrt(5)))) / (-sqrt(2 + 2/sqrt(5)))).simplify() == 1
+    assert (sech(asech(sqrt(2*(2 + sqrt(2))))) / (sqrt(2*(2 + sqrt(2))))).simplify() == 1
+    assert (sech(asech((1 + sqrt(5)))) / ((1 + sqrt(5)))).simplify() == 1
+    assert (sech(asech((-1 - sqrt(5)))) / ((-1 - sqrt(5)))).simplify() == 1
+    assert ((sech(asech((-sqrt(6) - sqrt(2))))) / ((-sqrt(6) - sqrt(2)))).simplify() == 1
```
Comment (jksuom) [inline sympy/functions/elementary/tests/test_hyperbolic.py]: There seems to be an extra pair of parentheses.

**Comment by jksuom** [comment]
This looks quite good. The simplification could be further 'simplified':
```
In [6]: expand_mul(sech(asech(sqrt(6) - sqrt(2))) / (sqrt(6) - sqrt(2)))
Out[6]: 1
```

**Comment by aktech** [comment]
I think `simplify()` looks more intuitive and readable & moreover `expand_mul` doesn't works in all the cases.

```python
In [2]: expand_mul(sech(asech(sqrt(2 + 2/sqrt(5)))) / (sqrt(2 + 2/sqrt(5))))
Out[2]: 1/(sqrt(-sqrt(5)/8 + 5/8)*sqrt(2*sqrt(5)/5 + 2))

In [3]: expand_mul(sech(asech(-sqrt(2 + 2/sqrt(5)))) / (-sqrt(2 + 2/sqrt(5))))
Out[3]: 1/(sqrt(-sqrt(5)/8 + 5/8)*sqrt(2*sqrt(5)/5 + 2))

In [4]: expand_mul(sech(asech(sqrt(2*(2 + sqrt(2))))) / (sqrt(2*(2 + sqrt(2)))))
Out[4]: 1/(sqrt(-sqrt(2)/4 + 1/2)*sqrt(2*sqrt(2) + 4))
```

**Comment by jksuom** [comment]
Perhaps those cases could be treated differently. I don't think the tests should necessarily look as intuitive as possible. Rather, one should strive for the shortest execution.

**Comment by aktech** [comment]
Shall I use `simplify()` for those 3 cases?

**Comment by aktech** [comment]
@jksuom 
What do you think now?

**Comment by jksuom** [comment]
I think it is ok to use simplify in those cases where no other (not too hard to find) method seems to work.

**Comment by aktech** [comment]
OK, I have pushed the changes.

**Comment by jksuom** [comment]
None, at least at the moment. I shall look into final forms still more closely. If nothing comes forth, you could squash the commit.

**Comment by aktech** [comment]
OK, Sure.

**Comment by aktech** [comment]
BTW, Shall I squash all commits to 1?

**Comment by jksuom** [comment]
Looks ok to me now. Would there be any reasons for not squashing all commits to 1?

**Comment by aktech** [comment]
> Would there be any reasons for not squashing all commits to 1?

None. I was just confirming.

**Comment by debugger22** [comment]
Recently I saw a repository which calls itself a style guide for git. I pretty much agree with it. See [this](https://github.com/agis-/git-style-guide#commits).

**Comment by asmeurer** [comment]
Another option is numeric testing. I think there are some utilities functions for helping with it.

**Comment by jksuom** [comment]
Merging in 24 hours.

**Comment by aktech** [comment]
@asmeurer 
I have had added a couple of tests for  numeric testing [here](https://github.com/sympy/sympy/pull/9291/files#diff-8b51859e2fcf6cf462fcbc0557f45908R554).

@debugger22 
Thanks for the link, looks good.
